[Totsuka Tomofumi] iP

totsukatomofumi opened the pull request:
### DukePro

> "Your mind is for having ideas, not holding them." – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/totsukatomofumi/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
##
If you Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Duke {
  public static void main(String[] args) {
    new Duke("./data/history.txt").run();
  }
}
```

Commit:
Add Gradle support

totsukatomofumi committed:
Git to ignore temp swap files.

totsukatomofumi committed:
Level 1. Greet, Echo, Exit

totsukatomofumi committed:
Level 2. Add, List

totsukatomofumi committed:
Level 3. Mark as Done

totsukatomofumi committed:
Level 4. ToDos, Events, Deadlines
Level 5. Handle Errors

totsukatomofumi committed:
Level 6. Delete

totsukatomofumi committed:
A-TextUiTesting: Automated Text UI Testing

totsukatomofumi committed:
Add function to save task history to a text file

The Level-7 increment requires me to add the function of saving
a session's task history to a text file located in a directory.

Let's,
* make that function a reality
* change the JavaDoc comments slightly

totsukatomofumi committed:
Add function to parse dates and convert between date formats

Increment Level-8 requires us to implement a way for Duke to handle
dates. We can go further and enable Duke to handle times as well.

Let's implement a way for Duke to parse, store and format convert
dates.

For now, handling times is not implemented but will try to in the
future.

totsukatomofumi committed:
Follow iP requirements to merge without fast-forward

Merge branch 'branch-Level-7'

totsukatomofumi committed:
Merge branch 'branch-Level-8'

totsukatomofumi committed:
Fix the history storing function that was broken by merging

totsukatomofumi committed:
Use more classes to achieve better OOP

Increment A-MoreOOP requires us to use more classes for better OOP.
Creating more classes will be better OOP where different classes
have their own roles.

Let's create more classes to split the logic into their each roles.
Let's also clean up the code and the logic to become neater.

totsukatomofumi committed:
Refactor package structure and package name

Increment A-Packages requires us to divide the project into packages
and subpackages. Also, task list stored in text file sometimes misses
last task added during the session.

Let's,
* divide the project into packages and subpackages
* fix the task history issue

totsukatomofumi committed:
Merge branch to add Gradle wrapper files to project

totsukatomofumi committed:
Change directory structure to get Gradle to work

totsukatomofumi committed:
Change gradle config

totsukatomofumi committed:
TaskList.java: Fix usage of raw type

totsukatomofumi committed:
build.grade: Add deletion of files made by tests to clean task

totsukatomofumi committed:
TaskList.java: Change access modifier of retrieve() to public

totsukatomofumi committed:
Add JUnit tests to test the behavior of the code

This increment requires us to implement more than two test methods,
preferably targeting more than one class.

Let's implement tests for classes containing complex logic.

totsukatomofumi committed:
Add JavaDoc to all java files

totsukatomofumi committed:
Tweak to comply with the coding standard

totsukatomofumi committed:
Refactor package structure of test code files

totsukatomofumi committed:
.getignore: Ignore files made in testing

totsukatomofumi committed:
TaskListTest.java: Remove unused import statement

totsukatomofumi committed:
Add command to find keyword from existing tasks

totsukatomofumi committed:
Add override annotation to overriding methods

totsukatomofumi committed:
build.gradle: Add deletion of files made by Duke to clean task

totsukatomofumi committed:
Merge branch 'branch-A-JavaDoc'

totsukatomofumi committed:
Merge tag 'A-CodingStandard'

A-CodingStandard

totsukatomofumi committed:
Merge branch 'branch-Level-9'

ALWHgithub commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,24 @@
+     */
+    @Override
+    public void run() {
+        System.out.println("Bye. Hope to see you again soon!");
```
Should the print statements be moved to the UI class, since they deal with the user ?
Noticed the same thing with the other commands

ALWHgithub commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,53 @@
+    public void run() {
+        taskList.add(new Deadline(description, time));
+        System.out.println("Got it. I've added this task:");
+        System.out.println(taskList.get(taskList.size() - 1).toString());
```
Should getting the size of the taskList be encapsulated to the TaskList class instead ?

ALWHgithub commented on src/main/java/duke/exception/IllegalDescriptionException.java:
```diff
@@ -0,0 +1,16 @@
+package duke.exception;
+
+/**
+ * IllegalDescriptionException is an exception relating to invalid description input.
+ *
+ * @author totsukatomofumi
+ */
+public class IllegalDescriptionException extends IllegalArgumentException {
+    /**
+     * Constructs an IllegalDescriptionException with the specified detail message.
+     * @param s the detail message.
+     */
+    public IllegalDescriptionException(String s) {
+        super(s);
+    }
+}
```
You may want to consider making a single DukeException for all your exceptions, since all of your exceptions are essentially the same

Throwing of the DukeException could be done inside the if-else statement (with different messages for different errors)

ALWHgithub commented on src/main/java/duke/logic/Parser.java:
```diff
@@ -0,0 +1,275 @@
+        }
+        //convert month to number
+        if (sMonth.length() == 3) {  //length of 3 may indicate abbrev
+            sMonth = Parser.months.get(sMonth.toLowerCase());
```
I like that you checked the month manually, but maybe you could use the LocalDate's native parsing function to simplify the code a bit.

aaronliwh reviewed (commented):
Generally good coding standard, just a few minor things :)

aaronliwh commented on src/main/java/duke/logic/Parser.java:
```diff
@@ -0,0 +1,275 @@
+package duke.logic;
+
+import duke.command.*;
+import duke.exception.*;
```
Should this be expanded out to explicit imports?
I noticed this in a few other places too

totsukatomofumi replied:
I am not opposed to the use of explicit imports. However, as the class will be using the entire `duke.command` subpackage, I have decided it is much cleaner to use wildcard import.

aaronliwh commented on src/main/java/duke/logic/TaskList.java:
```diff
@@ -0,0 +1,196 @@
+     * @param query the zero-based index.
+     * @return true if a task exists, else false.
+     */
+    public boolean exists(int query) {
```
Should functions be named this way ? Consider getExists.

aaronliwh commented on src/main/java/duke/logic/TaskList.java:
```diff
@@ -0,0 +1,196 @@
+     */
+    @Override
+    public boolean add(Task task) {
+        boolean bool = super.add(task);
```
Should variables be named this way? Consider a more descriptive name like taskAdded

aaronliwh reviewed (commented):
Generally good coding standard, just a few minor things :)

RussellDash332 commented on src/main/java/duke/logic/Parser.java:
```diff
@@ -0,0 +1,275 @@
+    private TaskList taskList;
+
+    /** HashMap to contain abbreviated month and their numerical representation as a string pairs. */
+    private static HashMap<String, String> months = new HashMap<>();
+
+    static {
+        String[] monthsArr = {"jan", "feb", "mar", "apr", "may", "jun", "jul", "aug", "sep", "oct", "nov", "dec"};
```
Since the string array is named `monthsArr`, would it better for the hashmap to be named `monthsMap`?

RussellDash332 commented on src/main/java/duke/logic/Parser.java:
```diff
@@ -0,0 +1,275 @@
+    public Runnable parse(String response) {
+        //strip() to allow for any (unintentional) whitespaces before or after
+        response = response.strip();
+        String lresponse = response.toLowerCase();  //for any caps commands
```
Should `lresponse` be renamed into something that isn't abbreviated such as `lowercaseResponse`? The `l` could've meant left instead of lowercase.

RussellDash332 commented on src/main/java/duke/logic/Parser.java:
```diff
@@ -0,0 +1,275 @@
+        //ddmmyyyy
+        String[] arr = new String[3];
+        //forward slash, hyphen, dot, or whitespace, otherwise no separator cases
+        if (str.indexOf('/') >= 0) {
+            arr = Parser.parseTimeHelper('/', str);
+        } else if (str.indexOf('-') >= 0) {
+            arr = Parser.parseTimeHelper('-', str);
+        } else if (str.indexOf('.') >= 0) {
+            arr = Parser.parseTimeHelper('.', str);
+        } else if (str.indexOf(' ') >= 0) {
+            arr = Parser.parseTimeHelper(' ', str);
+        } else if (str.length() == 4 || str.length() == 6 || str.length() == 8) {
```
Would `parseTimeDelimited` be a better naming since `Helper` sounds more general so it conveys less detail on what it actually does?

RussellDash332 commented on src/main/java/duke/logic/TaskList.java:
```diff
@@ -0,0 +1,196 @@
+     *
+     * @throws IOException If the file contains invalid contents that cannot be parsed.
+     */
+    public void retrieve() throws IOException {
```
What do you think of a more verbose naming such as `retrieveListOfTasks`?

RussellDash332 commented on src/test/java/duke/logic/ParserTest.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    @Test
+    public void parse_badInput_printMessage() {
```
Good test case naming! This will make readers understand what the testcase does very quickly 👍🏻

RussellDash332 reviewed (commented):
LGTM! Considerably few nits to deal with, but overall I like how flawless almost all of your namings are 🚀 🚀 

Overall, a very good effort on the iP, I'd love to see what's to come 😁

totsukatomofumi committed:
Add checkstyle to project

totsukatomofumi committed:
Implement graphic user interface

Will do in future:
* improve user interface (chat bubbles, separating responses into
  separate bubbles, etc.)
* update javadoc
* update tests

totsukatomofumi committed:
Add tweaks to the graphic user interface

totsukatomofumi committed:
Merge branch 'branch-A-CheckStyle'

totsukatomofumi committed:
Merge branch 'branch-Level-10'

totsukatomofumi committed:
MainWindow.fxml: Fix textfield color for easier visibility

totsukatomofumi committed:
Improve code quality

Code quality could be improved. Method and variable names could be
more descriptive. Long methods could be extracted out of.

By improving quality of code, readers will not find it hard to
understand the code.

Let's,
* rename methods
* extract methods out of long methods in Parser and TaskList
  classes
* move method from TaskList to Storage classes

Methods will be much more descriptive for the reader. Extracting
from long methods will allow for much more readability.
Method was moved in consideration of better heeding the OOP
paradigm.

totsukatomofumi committed:
Add use of assertions

Assertions help to check if important assumptions hold at certain
points of the code. This ensures that the program will terminate
before any unintended consequence.

Let's add assertions into the code where important assumptions
were made.

So that important assumptions are holding.

totsukatomofumi committed:
Merge pull request #2 from totsukatomofumi/branch-A-Assertions

Add use of assertions

totsukatomofumi committed:
Merge branch 'master' of https://github.com/totsukatomofumi/ip

totsukatomofumi committed:
Merge branch 'master' into branch-A-CodeQuality

totsukatomofumi committed:
Merge pull request #3 from totsukatomofumi/branch-A-CodeQuality

Improve code quality

totsukatomofumi committed:
Merge branch 'master' of https://github.com/totsukatomofumi/ip

totsukatomofumi committed:
Storage class: add an assert statement

totsukatomofumi committed:
Commands: fix bugs caused by bad conflict resolution

totsukatomofumi committed:
Storage class: fix a history retrieval issue

totsukatomofumi committed:
Add the ability to recognize and warn of duplicate tasks

totsukatomofumi committed:
Merge branch 'branch-C-DetectDuplicates'

totsukatomofumi committed:
build.grade: change application settings

totsukatomofumi committed:
Junit tests: fix what was broken from UI to GUI change

totsukatomofumi committed:
Task class: change contains behaviour to be case insensitive

totsukatomofumi committed:
Update docs files for release
